server/src/main/java/org/hyperskill/community/flashcards/card/CardController.java

```diff
@@ -39,4 +50,39 @@ public long getCardsCount(@RequestParam(name = "categoryId") String categoryId)
         var username = authenticationResolver.resolveUsername();
         return cardService.getCardCountOfCategory(username, categoryId);
     }
+
+    @GetMapping("/{cardId}")
+    public CardResponse getCardById(@PathVariable String cardId, @RequestParam String categoryId) {
+        var username = authenticationResolver.resolveUsername();
+        var card = cardService.getCardById(username, cardId, categoryId);
+        return mapper.map(card, categoryId);
+    }
+
+    @DeleteMapping("/{cardId}")
+    public void deleteCardById(@PathVariable String cardId, @RequestParam String categoryId) {
+        var username = authenticationResolver.resolveUsername();
+        cardService.deleteCardById(username, cardId, categoryId);
+    }
+
+    @PostMapping
+    public ResponseEntity<Void> createCard(
+            @RequestParam String categoryId,
+            @Valid @RequestBody CardCreateRequest request
+    ) {
+        var username = authenticationResolver.resolveUsername();
+        var id = cardService.createCard(username, request, categoryId);
+        var uri = URI.create("/api/cards/" + id);
+        return ResponseEntity.created(uri).build();
+    }
+
+    @PutMapping("/{cardId}")
+    public CardResponse updateCardById(
+            @PathVariable String cardId,
+            @RequestParam String categoryId,
+            @Valid @RequestBody CardUpdateRequest request
+    ) {
+        var username = authenticationResolver.resolveUsername();
+        var card = cardService.updateCardById(username, cardId, request, categoryId);
+        return mapper.map(card, categoryId);
```

Comment:
the create and update controller methods provide the request to the server, which is mapped to the document there. On the other hand you provide the documents for get to the controller and map them here.
If we map request to document already in the controller, the mapper is only needed to autowire here - and the request / response dto dependencies stay only in the web (controller) layer.

Not a crucial thing - but I'd find that a bit cleaner architectural..

Reply:
will create an issue. approve for now